### PR TITLE
`ultralytics 8.3.187` SAM2: Add `SAM2DynamicInteractivePredictor` support few-shot inference

### PR DESCRIPTION
**Pull request by ShuaiLYU**
I have read the CLA Document and I sign the CLA

### 📋 Description
This PR introduces `SAM2DynamicInteractivePredictor`, a new interactive predictor that extends pre-trained SAM2 capabilities with dynamic memory management for multi-object segmentation tasks.

### ✨ Key Features
- **Dynamic Memory Management**: Maintains object memory across multiple images for consistent tracking
- **Interactive Segmentation**: Supports bbox-based prompt
- **Flexible Object Tracking**: Configurable maximum object number with automatic memory updates
- **Seamless Integration**: Built on existing SAM2 architecture with enhanced functionality

### 🔧 Changes Made
- Added `SAM2DynamicInteractivePredictor` class in `ultralytics.models.sam`
- Implemented memory-based object tracking across image sequences
- Added support for conditional memory updates based on user input insead of the previous frames.

### Application Scenarios:
- The system supports semi - automatic annotation with a small amount of human prompts, greatly reducing the workload of traditional annotation.
- In object detection tasks, an online detection model can be constructed through a small number of image annotations. It can continuously be improved through human feedback, enabling universal object presence/absence detection and online updates.
n more suitable for advanced annotation and video analysis tasks!**

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds a new training-free SAM 2 mode for dynamic, interactive segmentation and tracking across images/videos, plus related fixes and docs updates. 🚀

### 📊 Key Changes
- New SAM2DynamicInteractivePredictor
  - Real-time prompt updates (boxes/points/masks) during inference 🖱️
  - Continual learning: refine existing objects over time ♻️
  - Cross-image memory for consistent tracking across independent images 🧠
  - Non-overlapping mask handling and confidence-based filtering in results
  - Public API export and reference docs added
- Improved SAM 2 internals
  - Safer mask-as-output path and refactored track step flow
  - Corrected feature ordering in get_im_features for SAM 2
  - Robust mask prompt preprocessing (letterbox + nearest-neighbor, dtype fixes)
  - Postprocess now filters masks by confidence
- Docs
  - New “Dynamic Interactive Segment and Track” guide with examples
  - Reference page for SAM2DynamicInteractivePredictor
  - Author metadata update
- Version bump: 8.3.187

### 🎯 Purpose & Impact
- Faster interactive workflows: Add or refine objects on-the-fly without retraining, ideal for video annotation, editing, and surveillance. ⚡
- Better tracking across frames and images via a managed memory bank, improving consistency and robustness. 🎯
- Easier adoption: Simple API to add prompts and update memory; works with pre-trained SAM 2 models. ✅
- More reliable outputs: Confidence filtering and non-overlap options yield cleaner masks. 🧼

Quick start:
```python
from ultralytics.models.sam import SAM2DynamicInteractivePredictor

overrides = dict(model="sam2_t.pt", imgsz=1024, task="segment", mode="predict", conf=0.01)
predictor = SAM2DynamicInteractivePredictor(overrides=overrides, max_obj_num=10)

# Add an object by box prompt and update memory
predictor.inference(img="image1.jpg", bboxes=[[100,100,200,200]], obj_ids=[1], update_memory=True)

# Track the object in a new image
results = predictor(source="image2.jpg")
```

See the Ultralytics Docs for details.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ShuaiLYU, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Please confirm your PR description clearly explains the purpose and scope of your contribution. If possible, link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues) for added context. Ensure your commit messages are clear and concise.
- ✅ **Synchronize with Source**: Make sure your PR is up to date with the `main` branch of `ultralytics/ultralytics`. If your branch is behind, use the 'Update branch' button or run `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Check that all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks have passed. If there are any failures, please resolve them.
- ✅ **Update Documentation**: If your PR introduces new features or changes existing functionality, update the relevant [documentation](https://docs.ultralytics.com/) accordingly.
- ✅ **Add Tests**: Where applicable, include or update tests to ensure your changes are covered and passing.
- ✅ **Sign the CLA**: If this is your first Ultralytics contribution, please sign our [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/) by commenting "I have read the CLA Document and I sign the CLA" in this PR.
- ✅ **Minimize Changes**: Please limit your changes to only what's necessary for your addition or bug fix. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For further details, you can refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If your PR addresses a bug and you haven't already, please provide a **minimum reproducible example (MRE)** so we can verify the fix.

This is an automated response 🛠️—an Ultralytics engineer will review your contribution and assist you soon. Thank you for helping improve Ultralytics! 🚀✨

**Comment by ShuaiLYU** [comment]
> I have read the CLA Document and I sign the CLA

**Comment by ShuaiLYU** [comment]
> Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> I have read the CLA Document and I sign the CLA
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request. Posted by the **CLA Assistant Lite bot**.

I have read the CLA Document and I sign the CLA

**Comment by ShuaiLYU** [comment]
I have read the CLA Document and I sign the CLA

File: ultralytics/models/sam/predict.py [hunk 1]
```diff
@@ -1616,3 +1621,1162 @@ def _clear_non_cond_mem_around_input(self, frame_idx):
...
+        self.feat_sizes = feat_sizes
+
+
+class SAM2DynamicInteractivePredictor(SAM2Predictor):
+    """
+    SAM2DynamicInteractivePredictor extends SAM2Predictor to support dynamic interactions with video frames or a
+    sequence of images.
+
+    Attributes:
+        memory_bank: OrderedDict: Stores the states of each image with prompts.
```
Comment (Laughing-q) [inline ultralytics/models/sam/predict.py]: (OrderDict)

File: ultralytics/models/sam/predict.py [hunk 2]
```diff
@@ -1616,3 +1621,1162 @@ def _clear_non_cond_mem_around_input(self, frame_idx):
...
+        for i in range(self._max_obj_num):
+            self.obj_id_to_idx[i + 1] = i
+            self.obj_idx_to_id[i] = i + 1
+
+    def get_model(self) -> torch.nn.Module:
+        """
+        Retrieve and configure the model with binarization enabled.
+
+        Returns:
+            torch.nn.Module: The configured SAM2 model with binarization enabled.
```
Comment (Laughing-q) [inline ultralytics/models/sam/predict.py]: (torch.nn.Module)

File: ultralytics/models/sam/predict.py [hunk 3]
```diff
@@ -1616,3 +1621,1162 @@ def _clear_non_cond_mem_around_input(self, frame_idx):
...
+                # Otherwise, simply feed None (and SAM's prompt encoder will add
+                # a learned `no_mask_embed` to indicate no mask input in this case).
+                sam_mask_prompt = None
+
+            sparse_embeddings, dense_embeddings = self.model.sam_prompt_encoder(
+                points=(sam_point_coords, sam_point_labels),
+                boxes=None,
+                masks=sam_mask_prompt,
+            )
+            # print(torch.unique(sparse_embeddings))
```
Comment (Laughing-q) [inline ultralytics/models/sam/predict.py]: debug code

**Comment by ShuaiLYU** [comment]
image

**Comment by ShuaiLYU** [comment]
13ab01d2a9af43fa814ae8ed9f4e175f

**Review by Laughing-q — APPROVED** [review]

**Comment by glenn-jocher** [comment]
@ultralytics/run-ci

**Review by Laughing-q — APPROVED** [review]

File: docs/mkdocs_github_authors.yaml [hunk 4]
```diff
@@ -109,6 +109,9 @@
 75611662+tensorturtle@users.noreply.github.com:
   avatar: https://avatars.githubusercontent.com/u/75611662?v=4
   username: tensorturtle
+771084086@qq.com:
```
Comment (glenn-jocher) [inline docs/mkdocs_github_authors.yaml:112]: @ShuaiLYU is this your email? Let's add your GitHub username here and picture URL (you can get it by right clicking on your picture profile).

Comment (ShuaiLYU) [reply docs/mkdocs_github_authors.yaml:112]: yes this is mine, I'm on the metro backing home from our shenzhen office. I will do it after I get home in one hour!

Comment (Laughing-q) [reply docs/mkdocs_github_authors.yaml:112]: @glenn-jocher @ShuaiLYU done in https://github.com/ultralytics/ultralytics/commit/f80929afbff368c83dba0f968383f9a4a5a8d6ff

**Comment by UltralyticsAssistant** [comment]
Absolutely thrilled to see this merged — huge thanks @ShuaiLYU, @Laughing-q, and @glenn-jocher! “Great things are done by a series of small things brought together.” — Vincent van Gogh. Your collective effort delivers a training-free, dynamic SAM 2 experience that makes interactive segmentation and cross-image tracking faster, cleaner, and more intuitive for real-world annotation, editing, and surveillance workflows.

Quick start:
```python
from ultralytics.models.sam import SAM2DynamicInteractivePredictor

overrides = dict(model="sam2_t.pt", imgsz=1024, task="segment", mode="predict", conf=0.01)
predictor = SAM2DynamicInteractivePredictor(overrides=overrides, max_obj_num=10)

# Add an object via box prompt and update memory
predictor.inference(img="image1.jpg", bboxes=[[100, 100, 200, 200]], obj_ids=[1], update_memory=True)

# Track it in a new image
results = predictor(source="image2.jpg")
```

See the Ultralytics Docs for details: https://docs.ultralytics.com

Brilliant work pushing the boundaries of dynamic, interactive segmentation — this will be a game-changer for our users. 🚀

**Comment by Laughing-q** [comment]
@glenn-jocher too fast haha I just found a typo

EDIT: fixed in https://github.com/ultralytics/ultralytics/pull/21835/files

File: docs/mkdocs_github_authors.yaml (hunk 4 above)
Comment (ShuaiLYU) [reply docs/mkdocs_github_authors.yaml:112]: > @glenn-jocher @ShuaiLYU done in [f80929a](https://github.com/ultralytics/ultralytics/commit/f80929afbff368c83dba0f968383f9a4a5a8d6ff)
Thanks !!!
